### DISCUSSION TITLE
Ability to disable adoptStyles call

### DISCUSSION THREAD

**Discussion by pthomasius** [lit/lit#2412] [Ideas]
Hi,

We have a web app that runs in Openfin/Electron and uses their api to open new windows. The new window has access to window.opener and with that all JS is running in the main window. We than append a lit component created in the main window to the body of the new window. The adoptStyles call throws an exception as constructed stylesheets can not be shared across documents. There seems to be no way of stopping lit calling adoptStyles even if no super is called in the createRenderRoot override.
We already have code to load the styles inline for this scenario as it was already necessary before using lit. If we could stop the adoptStyles call we could use static styles in the main window and get the performance advantage of constructed stylesheets while loading inline styles in the new windows.
It seems currently the only way to make this work is by having all styles rendered inline and not use the lit static styles, which will bloat the DOM in the main window unnecessarily.

Any ideas how to conditionally disable adoptStyles?

Thanks for any ideas/comments

**Comment by justinfagnani** [comment]
I would generally avoid adopting components into different windows. In the future other things in the DOM may not work correctly, and components themselves may make assumptions about the window or document they were created in. Loading the component definitions you need in the new window should be the safest thing.

But if you want to disable adopted stylesheets globally, you can do this before any other scripts load:

```html
<head>
  <script>
    delete Document.prototype.adoptedStyleSheets;
    delete CSSStyleSheet.prototype.replace;
  </script>
```

**Comment by pthomasius** [comment]
Hi Justin,

Let me give some more context. This is a realtime financial application with a complex ui and streaming data. Using the well established document.adoptNode allows for the state of the view to be transferred. Re initialising the components in the new window and recreate all of the state is not a solution, in fact it is not possible as some of that state are streams to specific server instances that can't be recreated and would impose either a certain backend architecture or complex frontend logic were the component does something different when loaded into a detached window to hook into the existing streams on the main window to achieve this. Yes thirdparty libs making assumptions about document and window is a problem, we write our own ui components to ensure we use the element.ownerDocument and its defaultView for window. In fact you could argue that the adoptStyles call is making assumptions about the document as it will always try to assign the cached constructed stylesheet without a check and with it "break" document.adoptNode.
The solution given does not address the overhead of inline css in the main window.
I appreciate that lit is trying to keep it simple and do the right thing for the developers, I write the framework that the above applications are based on and know the balancing act. And the current implementation is certainly a good default, but it would be nice to have the ability to hook into the styles life cycle to accommodate these more complex scenarios. At a minimum it would mean the main window with most of the UI has the perf gain of adoptedStyles and the detached views use inline styles. Ideally it would allow for the constructed stylesheets to be created in the new window against the new document from the cssText and then assigned to adoptedStyleSheets in the adoptStyles call.

I'm currently navigating and debugging lit to understand more about this and just wanted to kick off a discussion. I'd be happy to have a call where I can show the app, as it is not public, for even more context.

Kind Regards
Pierre

**Reply by pthomasius** [reply]
After some more investigation it looks like it can be done in the createRenderRoot override after all. In earlier testing it seemed the super was always called regardless, but am not seeing that now. So I can conditionally adopt the styles using the exported lit function or add a styles tag filled from cssText just as the internal function would. This would allow for the minimum scenario I mentioned above and give me a hook to see if it is possible to create constructed stylesheets against the new document.

**Reply by pthomasius** [reply]
I've come up with a base class that uses the standard lit way in the context of the main window and creates/caches constructed stylesheets with the correct document context for the child windows.
If any one else is running into this problem.

`import { LitElement, ReactiveElement, CSSResult, adoptStyles } from "lit"

export class LitComponentBase extends LitElement {

    constructor() {
        super();
    }

    //this override is to ensure lit doesn't call adoptStyles 
    protected createRenderRoot(): Element | ShadowRoot {
        const renderRoot =
          this.shadowRoot ??
          this.attachShadow(
            (this.constructor as typeof ReactiveElement).shadowRootOptions
          );
        return renderRoot;
    }

    public connectedCallback(): void {
        super.connectedCallback();
        //apply styles when component is connected
        if(this.shadowRoot) {
            this.applyStyles(this.shadowRoot);
        }
    }

    private applyStyles(renderRoot): void {
        const styles = (this.constructor as typeof ReactiveElement).elementStyles;
        if(styles) {
            //if in the context of the main window call lit adoptStyles with static styles
            //same as lit would do
            //else call own create/cache in context of owner window and call adoptStyles with results
            if(this.ownerDocument === document){
                adoptStyles(renderRoot, styles);
            } else {
                const stylesheets: CSSStyleSheet[] = styles.map(s => this.createOrGetConstructedStyleSheet
                    .apply(this.ownerDocument.defaultView, [(s as CSSResult).cssText]));
                adoptStyles(renderRoot, stylesheets);
            }
        }
    }

    private createOrGetConstructedStyleSheet(cssText:string): {styleSheet:CSSStyleSheet} {
        //get the CSSStyleSheet constructor of the window passed as scope
        //as the default "new CSSStyleSheet()" will always execute in scope of the main window
        //and ref its document
        const styleSheetConstructor = this["CSSStyleSheet"];
        //create style cache if needed
        let styleSheetCache = this["__MY__styleSheetCache"];
        if(!styleSheetCache) {
            styleSheetCache = new Map<string, CSSStyleSheet>();
            this["__MY__styleSheetCache"] = styleSheetCache;
        }
        //create CSSStyleSheet with above constructor for correct document context
        //or return cached object
        let styleSheet = styleSheetCache.get(cssText);
        if(styleSheet === undefined) {
            styleSheetCache.set(cssText, (styleSheet = new styleSheetConstructor()));
            styleSheet.replaceSync(cssText);
        }
        //the result needs to be wrapped as the "instance of" check fails in lit adoptStyles
        //this mimics a CSSResult
        return {styleSheet: styleSheet};
    }
}`
